src/python/pants/engine/native.py

```diff
@@ -706,9 +716,26 @@ def func(constraint):
       return Function(self.context.to_key(constraint))
     def tc(constraint):
       return TypeConstraint(self.context.to_key(constraint))
+    def flatten_type_map_to_constraints(type_constraints_map):
```

Comment:
I think that part of the elegance of the implementation you suggested is that it shouldn't be necessary for the native engine to know about the union (yet... we could decide to do that later)...

If at `RuleIndex` (or perhaps `Scheduler`?) creation time you:
```
# expand a Get for a Union type...
Get(TypeX, Union)

# ...into Gets for all of its members:
Get(TypeX, UnionMember1), Get(TypeX, UnionMember2), .. #etc
```
... then there is no need for the engine itself to know about the existence of the union.

Reply:
This is really good.

Reply:
I'm actually thinking now that we don't need to do any expansion, but rather just check all `Get`s for union types (I believe we can/should do this at `RuleIndex` creation time) are valid according to the registered union rules (not sure if this was your intended meaning). This gives us the guarantee you've described below of knowing that `Get`s are valid statically, which allows the engine to just generate a `Get` from the runtime type as usual. And that should allow us to remove any knowledge of unions from the engine.

Reply:
> I believe we can/should do this at RuleIndex creation time

I don't know if we can assume that a RuleIndex always represents a closed world (right now). But we do know that we have a closed world at Scheduler creation time... so at Scheduler creation time you could rewrite the contents of the RuleIndex to replace union-Gets with union-member-Gets?

Reply:
After diving into the implementation of removing all the union knowledge from the engine, I think that sounds correct (since that's where we're already linking up the `Get`s for unions in this PR anyway).

Reply:
It was! It seems like all the engine work was purely to provide a more specific error message on failing to resolve a union somehow. I did also convert a couple `panic!`s into `throw()`s, which enabled the technique you describe in another comment (just making a `Get` and letting the engine error out if there's no path).